### PR TITLE
[AIR] Suppress Pandas Future Warnings in data conversion

### LINKED ISSUE #29270: Second, we get loads on future and early warning messages when executing `.match_batch()` on the Ray data. See the attached image
Created from #29267

### PR DESCRIPTION
**Pull request by amogkam**
Signed-off-by: Amog Kamsetty <amogkamsetty@yahoo.com>

Starting from Pandas 1.5, a FutureWarning is raised about `.iloc` and `.loc` always being in-place operations.

```
/Users/amog/dev/ray/python/ray/air/util/data_batch_conversion.py:221: FutureWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`
    df.loc[:, col_name] = TensorArray(col)
```

As seen in #29270, this can often spam the output.

However, we actually want in-place setting instead of the old behavior, so we suppress these warnings instead of reverting to the old behavior as recommended here: https://stackoverflow.com/a/74193599

## Related issue number
Closes https://github.com/ray-project/ray/issues/29270

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by amogkam** [comment]
there actually seem to be significant differences in behavior between `df.loc` API and directly setting the column. https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

@clarkzinzow do you have more context on this?

**Review by clarkzinzow — CHANGES_REQUESTED** [review]
We should only be using the `iloc` and `loc` APIs, since we nearly always want to be setting these in-place. The recommendation in that warning is only if you're wanting a copy to be returned, which we typically do _not_ want. If we need a copy, that should be done upstream from these mutations.

**Comment by amogkam** [comment]
@clarkzinzow can you clarify what the recommended change should be then?

From the Pandas changelog, `df.iloc` will no longer work as we are currently using it in future Pandas versions. https://pandas.pydata.org/docs/dev/whatsnew/v1.5.0.html#inplace-operation-when-setting-values-with-loc-and-iloc.

**Comment by clarkzinzow** [comment]
@amogkam When we want to mutate the underlying DataFrame instead of creating an entirely new copy (which is nearly always the case), we want to use `.iloc[]` and `.loc[]`, and AFAICT those are the semantics we want in these three cases.

File: python/ray/air/util/data_batch_conversion.py [hunk 1]
```diff
@@ -93,7 +93,7 @@ def convert_pandas_to_batch_type(
     elif type == DataType.NUMPY:
         if len(data.columns) == 1:
             # If just a single column, return as a single numpy array.
-            return data.iloc[:, 0].to_numpy()
+            return data[data.columns[0]].to_numpy()
```
Comment (clarkzinzow) [inline python/ray/air/util/data_batch_conversion.py:96]: This is just returning an rvalue, we're not mutating here, so `data.iloc[:, 0].to_numpy()` should be fine.

File: python/ray/air/util/data_batch_conversion.py [hunk 2]
```diff
@@ -218,7 +218,7 @@ def _cast_ndarray_columns_to_tensor_extension(df: pd.DataFrame) -> pd.DataFrame:
         for col_name, col in df.items():
             if column_needs_tensor_extension(col):
                 try:
-                    df.loc[:, col_name] = TensorArray(col)
+                    df[col_name] = TensorArray(col)
```
Comment (clarkzinzow) [inline python/ray/air/util/data_batch_conversion.py:221]: We want to mutate (overwrite) the column in-place, so `df.loc[:, col_name] = TensorArray(col)` is preferred.

File: python/ray/air/util/data_batch_conversion.py [hunk 3]
```diff
@@ -238,5 +238,5 @@ def _cast_tensor_columns_to_ndarrays(df: pd.DataFrame) -> pd.DataFrame:
         # Try to convert any tensor extension columns to ndarray columns.
         for col_name, col in df.items():
             if isinstance(col.dtype, TensorDtype):
-                df.loc[:, col_name] = pd.Series(list(col.to_numpy()))
+                df[col_name] = pd.Series(list(col.to_numpy()))
```
Comment (clarkzinzow) [inline python/ray/air/util/data_batch_conversion.py:241]: Same here, we want to mutate (overwrite) the column in-place, so `df.loc[:, col_name] = pd.Series(list(col.to_numpy()))` seems correct to me.

**Review by clarkzinzow — APPROVED** [review]
LGTM!

File: python/ray/air/util/data_batch_conversion.py [hunk 4]
```diff
@@ -218,7 +219,13 @@ def _cast_ndarray_columns_to_tensor_extension(df: pd.DataFrame) -> pd.DataFrame:
         for col_name, col in df.items():
             if column_needs_tensor_extension(col):
                 try:
-                    df.loc[:, col_name] = TensorArray(col)
+                    # Surpress Pandas warnings:
```
Comment (amogkam) [inline python/ray/air/util/data_batch_conversion.py:222]:
```suggestion
                    # Suppress Pandas warnings:
```

File: python/ray/air/util/data_batch_conversion.py [hunk 5]
```diff
@@ -238,5 +245,11 @@ def _cast_tensor_columns_to_ndarrays(df: pd.DataFrame) -> pd.DataFrame:
         # Try to convert any tensor extension columns to ndarray columns.
         for col_name, col in df.items():
             if isinstance(col.dtype, TensorDtype):
-                df.loc[:, col_name] = pd.Series(list(col.to_numpy()))
+                # Surpress Pandas warnings:
```
Comment (amogkam) [inline python/ray/air/util/data_batch_conversion.py:248]:
```suggestion
                # Suppress Pandas warnings:
```

**Pull request merged**